web/source/osk/banner.ts

```diff
@@ -0,0 +1,259 @@
+///<reference path="visualKeyboard.ts" />
+namespace com.keyman.osk {
+  // Base class for a banner above the keyboard in the OSK
+
+  export abstract class Banner {
+    private _Visible: boolean;
+    private _Enabled: boolean;
+    private _Height: number; // pixels
+    protected div: HTMLDivElement;
+
+    public DEFAULT_HEIGHT: number = 40; // pixels
+
+    public get height():number {
+      return this._Height;
+    }
+
+    public set height(height:number) {
+      this._Height = height;
+      if (this.div) {
+        let ds = this.div.style;
+        ds.height = (height > 0) ? height + 'px' : '0px';
+      }
+    }
+
+    public get visible():boolean {
+      return this._Visible;
+    }
+
+    public set visible(visible: boolean) {
+      this._Visible = visible;
+
+      if (this.div) {
+        let ds = this.div.style;
+        ds.display=(this._Visible) ? 'block': 'none';
+      }
+    }
+
+    public get enable():boolean {
+      return this._Enabled;
+    }
+
+    public set enable(enable: boolean) {
+      this._Enabled = enable;
+    }
+
+    public constructor(visible: boolean, enabled: boolean, height?: number) {
+      let keymanweb = com.keyman.singleton;
+      let util = keymanweb.util;
+
+      let d = util._CreateElement('div');
+      d.id = "keymanweb_banner_bar";
+      d.className = "kmw-banner-bar";
+      this.div = d;
+
+      this.height = (height >= 0) ? height : this.DEFAULT_HEIGHT;
```

Comment:
oskManager (line 1306) factors in the banner height for determining `maxHeight`.

A blank banner isn't visible, but having a positive height would cause bugs.

Reply:
Okay, that's disturbing me even more! Internally, we need to ensure if height=0 that also visible=false. The easiest way to do this is to link `_visible` to `height`, so something like `property visible get { return height >= 0 }` rather than having a separate `_visible` property. Obviously then a set visible would then just set the height to zero.

Reply:
I've made the following changes
* Removed `_visible` and only using height
* Added an `update()` method that updates the banner styling (height and display)